kroxylicious-filters/kroxylicious-authorization/src/main/java/io/kroxylicious/filter/authorization/AuthorizationFilter.java

```diff
@@ -0,0 +1,282 @@
+ * to the entities (e.g. topics and consumer groups) within the Kafka protocol.</p>
+ *
+ * <p>This class actually only implements a common mechanism, delegating the actual enforcement to {@link ApiEnforcement}
+ * instances on a per-API key basis.</p>
+ */
+public class AuthorizationFilter implements RequestFilter, ResponseFilter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AuthorizationFilter.class);
+
+    static Map<ApiKeys, ApiEnforcement<?, ?>> apiEnforcement = new EnumMap<>(ApiKeys.class);
+
+    static {
+        // This filter "fails closed", rejecting all (apikey, apiversions)-combinations which is doesn't understand
+        // because a new api or version could introduce a reference to some authorizable entity, like a topic,
+        // which would result in information disclosure because this filter was not applying authz.
+        apiEnforcement.put(ApiKeys.API_VERSIONS, new Passthrough<ApiVersionsRequestData, ApiVersionsResponseData>(0, 4) {
+            @Override
+            CompletionStage<ResponseFilterResult> onResponse(ResponseHeaderData header, ApiVersionsResponseData response, FilterContext context,
+                                                             AuthorizationFilter authorizationFilter) {
+                return authorizationFilter.checkCompat(header, response, context);
+            }
+        });
+        apiEnforcement.put(ApiKeys.SASL_HANDSHAKE, new Passthrough<>(0, 1));
+        apiEnforcement.put(ApiKeys.SASL_AUTHENTICATE, new Passthrough<>(0, 2));
+
+        apiEnforcement.put(ApiKeys.METADATA, new MetadataEnforcement());
+    }
+
+    @VisibleForTesting
+    public static boolean isApiSupported(ApiKeys apiKey) {
+        return apiEnforcement.containsKey(apiKey);
+    }
+
+    @VisibleForTesting
+    public static boolean isApiVersionSupported(ApiKeys apiKey, short apiVersion) {
+        var enforcement = apiEnforcement.get(apiKey);
+        if (enforcement == null) {
+            return false;
+        }
+        return enforcement.minSupportedVersion() <= apiVersion
+                && apiVersion <= enforcement.maxSupportedVersion();
+    }
+
+    @VisibleForTesting
+    public static short minSupportedApiVersion(ApiKeys apiKey) {
+        var enforcement = apiEnforcement.get(apiKey);
+        if (enforcement == null) {
+            return -1;
+        }
+        return enforcement.minSupportedVersion();
+    }
+
+    @VisibleForTesting
+    public static short maxSupportedApiVersion(ApiKeys apiKey) {
+        var enforcement = apiEnforcement.get(apiKey);
+        if (enforcement == null) {
+            return -1;
+        }
+        return enforcement.maxSupportedVersion();
+    }
+
+    // This filter need to inspect requests, filtering out unauthorized entities before forwarding to the broker.
+    // Then, when handling the response, we need to "merge" back the unauthorized entities we previously filtered out
```

Comment:
This surprised me.  I had assumed that Kafka would authorize requests first, and fail the whole request if any authorization error occured before any mutating operation was made.  Looking at KafkaApis.scala, I see am I wrong.